pkg/deploy/controller/deploymentconfig/controller.go

```diff
@@ -69,87 +70,89 @@ func (c *DeploymentConfigController) Handle(config *deployapi.DeploymentConfig)
-			if deployutil.DeploymentVersionFor(inflightDeployment) < deployutil.DeploymentVersionFor(&deployment) {
-				deploymentForCancellation, inflightDeployment = inflightDeployment, &deployment
-			} else {
-				deploymentForCancellation = &deployment
-			}
-
-			deploymentForCancellation.Annotations[deployapi.DeploymentCancelledAnnotation] = deployapi.DeploymentCancelledAnnotationValue
-			deploymentForCancellation.Annotations[deployapi.DeploymentStatusReasonAnnotation] = deployapi.DeploymentCancelledNewerDeploymentExists
-			if _, err := c.deploymentClient.updateDeployment(deploymentForCancellation.Namespace, deploymentForCancellation); err != nil {
-				util.HandleError(fmt.Errorf("couldn't cancel deployment %s: %v", deployutil.LabelForDeployment(deploymentForCancellation), err))
+			// Cancel running deployments.
+			inflightDeploymentsExist = true
+			if !deployutil.IsDeploymentCancelled(&deployment) {
+				deployment.Annotations[deployapi.DeploymentCancelledAnnotation] = deployapi.DeploymentCancelledAnnotationValue
+				deployment.Annotations[deployapi.DeploymentStatusReasonAnnotation] = deployapi.DeploymentCancelledNewerDeploymentExists
+				if _, err := c.deploymentClient.updateDeployment(deployment.Namespace, &deployment); err != nil {
+					util.HandleError(fmt.Errorf("couldn't cancel deployment %s: %v", deployutil.LabelForDeployment(&deployment), err))
+				}
+				glog.V(4).Infof("Cancelled deployment %s for deployment config %s", deployutil.LabelForDeployment(&deployment), deployutil.LabelForDeploymentConfig(config))
 			}
-			glog.V(4).Infof("Cancelled deployment %s for deployment config %s", deployutil.LabelForDeployment(deploymentForCancellation), deployutil.LabelForDeploymentConfig(config))
 		}
 	}
-
-	// if the latest deployment exists then nothing else needs to be done
-	latestDeploymentExists, _ := deployutil.LatestDeploymentInfo(config, existingDeployments)
-	if latestDeploymentExists {
-		return nil
-	}
-
-	// check to see if there are inflight deployments
-	if inflightDeployment != nil {
+	// Wait for deployment cancellations. We don't want to compete with
+	// deployment process scaling or lauch competing processes.
+	if inflightDeploymentsExist {
+		c.recorder.Eventf(config, "DeploymentAwaitingCancellation", "Deployment %d awaiting cancellation of running deployments", config.LatestVersion)
 		// raise a transientError so that the deployment config can be re-queued
-		glog.V(4).Infof("Found previous inflight deployment for %s - will requeue", deployutil.LabelForDeploymentConfig(config))
 		return transientError(fmt.Sprintf("found previous inflight deployment for %s - requeuing", deployutil.LabelForDeploymentConfig(config)))
 	}
-
-	// Try and build a deployment for the config.
+	// If the latest deployment already exists, reconcile replica counts.
+	if latestDeploymentExists {
+		// If the latest deployment exists and is still running, there's nothing
+		// to do.
```

Comment:
We could change the target replicas annotation here so that the rolling deployer shoots for the updated version